_posts/2019-02-14-Viewability.md

```diff
@@ -0,0 +1,115 @@
+
+So, now that I got this out of my system, I want to explain, how we measure and deliver viewability.
+
+### Measuring
+
+The viewability itself is measured in the adunit itself. Thus, JavaScript code is executed as part of the AdUnit. This code detects, if we can measure the viewability and if yes, it was viewable. Since we rely on the available APIs, the way of measurement depends strongly on the environment, e.g. mobile web vs. app and which version of APIs and browsers are available.
+
+#### App
+In apps, we first check if a version of [MRAID](https://www.iab.com/guidelines/mobile-rich-media-ad-interface-definitions-mraid/) is available. MRAID is a bridge of the native component of an app into the JavaScript container the adunit is running in. If MRAID is available, we can simply call
+
+```javascript
+mraid.isViewable()
+```
+
+to know, whether the is currently visible or not. But since we need not only to know, if the adunit is viewable, but also how long, we need to monitor the status. Instead of recalling the upper function every couple of milliseconds, we make use viewable change event, specified in MRAID:
+
+```javascript
+var mraidEvent = mraid.EventType ? (mraid.EventType.VIEWABLE_CHANGE || 'viewableChange') : 'viewableChange';
+mraid.addEventListener(mraidEvent, onMraidViewableChange);
+
+function onMraidViewableChange() {
+   mraidViewable = mraid.isViewable();
+}
+```
+
+
+#### Mobile web
+
+The mobile web comes with a bigger variety of environments. Safari and Chrome are the two main browsers and they come with different capabilites. In addition, the available information depends on the fact, if the adunit runs in a friendly or unfriendly iframe. An iframe is unfriendly if it is running with a different source than the original website.
+
+##### 1. Intersection Observer
+The simplest and most reliable case are browsers which support the [IntersectionObserver API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API). In this case we can just call it with
+
+```javascript
+    var _ioSettings = {
+       threshold: 0.5
+    };
+    var _ioCallback = function(entry) {
+       isElementVisible = entry[0].intersectionRatio >= 0.5;
+    };
+    var elementObserver = new IntersectionObserver(_ioCallback, _ioSettings);
+    elementObserver.observe(element);
+```
+
+
+##### 2. Safari painting behavior
+When this API is not available, e.g. in older browsers, we try different methods. If the ad is running in a modern Safari browser and in an unfriendly iframe, we can make use of an optimization mechanism of Safari. It prevents the rendering of content in unfriendly iframes if they are not in the viewport. Therefore, we are putting a beacon in the middle of the iframe and check if is rendering. If it is, we know that the ad is viewable.
```

Comment:
suggest "if they are not in the viewport" -> "if they are currently not in the viewport"